include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h

```diff
@@ -600,6 +600,66 @@ using __repacked_tuple_t = typename __repacked_tuple<T>::type;
 template <typename _ContainerOrIterable>
 using __value_t = typename __internal::__memobj_traits<_ContainerOrIterable>::value_type;
 
+//-----------------------------------------------------------------------
+// future and helper classes for async pattern/algorithm
+//-----------------------------------------------------------------------
+
+struct __tmp_base
+{
+};
+
+template <typename... Ts>
+struct __TempObjs : public __tmp_base
+{
+    ::std::tuple<Ts&...> __my_tmps;
+    __TempObjs(Ts&... __t) : __my_tmps(::std::forward_as_tuple(__t...)) {}
+};
+
+class __future_base
+{
+    using event = sycl::event;
+    event __my_event;
+
+  public:
+    __future_base(event __e) : __my_event(__e) {}
+    event
+    get_event() const
+    {
+        return __my_event;
+    }
+    void
+    wait()
+    {
+#if !ONEDPL_ALLOW_DEFERRED_WAITING
+        __my_event.wait();
+#endif
+    }
+    operator event() const { return event(__my_event); }
+};
+
+// TODO: Extend to support value type and sycl iterator.
+template <typename T>
+class __future : public __future_base
```

Comment:
When do we want to support value_type? in the upcoming release?

Reply:
It's only needed for non-void futures. I was planning to keep them separate for the next release.